pop_tools/zonal_average.py

```diff
@@ -0,0 +1,212 @@
+    """
+
+    # Allow user to input weights file, if there is not one, use default check
+    if weight_file is None:
+        weight_file = _get_default_filename(src_grid, dst_grid, method)
+
+    # Check to see if the weights file already exists - if not, generate weights
+    if not os.path.exists(weight_file) or clobber:
+        xe.Regridder(src_grid, dst_grid, method).to_netcdf(weight_file)
+
+    regridder = xe.Regridder(src_grid, dst_grid, method, weights=weight_file)
+
+    return regridder
+
+
+class regridder(object):
+    def __init__(
+        self,
+        grid_name,
+        grid=None,
+        dx=None,
+        dy=None,
+        mask=None,
+        regrid_method='conservative',
+        method_gen_grid='regular_lat_lon',
+    ):
+        """
+        A regridding class which uses xESMF and Xarray tools to both regrid and
+        calculate a zonal averge.
+
+        Parameters
+        ----------
+        grid_name
+        """
+        if grid_name is not None:
+            self.grid_name = grid_name
+
+            # Use pop-tools to retrieve the grid
+            self.grid = get_grid(grid_name)
+
+        elif grid is not None:
+            self.grid = grid
+
+        else:
+            raise ValueError('Failed to input grid name or grid dataset')
+
+        # Set the dx/dy parameters for generating the grid
+        self.dx = dx
+        self.dy = dy
+
+        # Set the regridding method
+        self.regrid_method = regrid_method
+
+        # Set the grid generation method
+        self.method_gen_grid = method_gen_grid
+
+        # If the user does not input a mask, use default mask
+        if not mask:
+            self.mask = self.grid['REGION_MASK']
+            self.mask_labels = self.grid['region_name']
+
+        else:
+            self.mask = mask
+
+    # Setup method for regridding a dataarray
+    def _regrid_dataarray(self, da_in, regrid_mask=False, regrid_method=None):
+
+        src_grid = _convert_to_xesmf(da_in, self.grid)
+        dst_grid = _generate_dest_grid(self.dy, self.dx, self.method_gen_grid)
+
+        # If the user does not specify a regridding method, use default conservative
+        if regrid_method is None:
+            regridder = _generate_weights(src_grid, dst_grid, self.regrid_method)
+
+        else:
+            regridder = _generate_weights(src_grid, dst_grid, regrid_method)
+
+        # Regrid the input data array, assigning the original attributes
+        da_out = regridder(src_grid[da_in.name])
+        da_out.attrs = da_in.attrs
+
+        return da_out
+
+    def regrid(self, obj, **kwargs):
+        """generic interface for regridding DataArray or Dataset"""
+        if isinstance(obj, xr.Dataset):
+            return obj.map(self._regrid_dataarray, keep_attrs=True, **kwargs)
+        elif isinstance(obj, xr.DataArray):
+            return self._regrid_dataarray(obj, **kwargs)
+        else:
+            raise ValueError('unknown type')
```

Comment:
The `else` isn't needed here because the only to only way to reach this branch is if every other condition hasn't been met

```suggestion
        raise ValueError('unknown type')
```